test/test_ops.py

```diff
@@ -399,7 +399,7 @@ def test_nms_cuda(self):
             r_cpu = ops.nms(boxes, scores, iou)
             r_cuda = ops.nms(boxes.cuda(), scores.cuda(), iou)
 
-            self.assertTrue(torch.allclose(r_cpu, r_cuda.cpu()), err_msg.format(iou))
+            self.assertTrue(set(r_cpu.tolist()) == set(r_cuda.tolist()), err_msg.format(iou))
```

Comment:
The returned order of the elements is kind of important -- we ensure that the indices are sorted wrt the scores in decreasing order.

Ideally, we would like to keep this test, but take into account the fact that if two elements have the same score, the order doesn't matter.
One way of doing this is by doing something like what [PyTorch does for testing `topk`](https://github.com/pytorch/pytorch/blob/945d7a7408891e25bc54a65015724f6d2de644e6/test/test_torch.py#L1625-L1630), roughly in the lines of:
```python
is_eq = torch.allclose(r_cpu, r_cuda.cpu())
if not is_eq:
    # if the indices are not the same, ensure that it's because the scores
    # are duplicate
    is_eq = torch.allclose(scores[r_cpu], scores[r_cuda.cpu()])
self.assertTrue(is_eq, err_msg.format(iou)
```

Reply:
This problem seems definitely due to lack of stability of torch's `tensor.sort()`, and that can be recreated with just a little snippet:

```
$ cat sorttest.py 
import torch

for s in range(150):
    torch.manual_seed(s)
    scores = torch.rand(1000)
    sort_indices_cpu = scores.sort(0, descending=True)[1]
    sort_indices_cuda = scores.cuda().sort(0, descending=True)[1]

    if not torch.allclose(sort_indices_cpu, sort_indices_cuda.cpu()):
        print(f'FAIL with seed {s}:')
        print(f'CPU CUDA')
        for a, b in zip(sort_indices_cpu.tolist(), sort_indices_cuda.tolist()):
            if a != b:
                print(f'{a:3} {b:3}')

$ python sorttest.py 
FAIL with seed 64:
CPU CUDA
810 478
478 810
FAIL with seed 122:
CPU CUDA
 58 427
427  58
```

Your suggested solution is much better; will update the PR.

Reply:
If a pair of same-scored boxes (A, B) happened to overlap beyond the IoU threshold, then one result would include box A, and the other box B if they sorted in different orders.

I guess a deeper solution for this would be to force the use of stable sort, but it's not clear that torch provides one. The docs for `torch.sort` don't mention stability either way.

Reply:
PyTorch doesn't enforce any guarantees on the order of the sorted elements (specially in CUDA). So I guess we will be taking the same approach here

Reply:
Can you revert this change, and we keep the other change? I think tests should pass in this case?

Reply:
Added a couple commits to:
- adjust the assertion as you suggested above
- tweak the data creation change to make it bit more straightforward (under the assumption that the intent is to create at least one box that will (barely) be suppressed)

Let me know what you think

Reply:
Oh, and looped over 5000 random seeds and those all passed.

Reply:
Ah, the root of this may be the difference in how the CUDA implementation (in `devIoU()`) and the CPU implementation (in `nms_cpu_kernel()`) determine whether the IoU for a box-pair is over threshold.

The `devIoU()` code multiplies total area by the threshold then compares to the intersection:

```
  return interS > threshold * (Sa + Sb - interS);
``

While `nms_cpu_kernel()` divides intersection by total and then compares to the threshold:

```
      auto ovr = inter / (iarea + areas[j] - inter);
      if (ovr > iou_threshold)
```

Those calculations give opposite results for some of the inputs that were being generated by the test. Here's a little C program that uses the guts of `devIoU()` to demonstrate:

```
$ cat iou.c   
#include <stdio.h>                           

#define MIN(x, y) (x) < (y) ? (x) : (y)      
#define MAX(x, y) (x) > (y) ? (x) : (y)      

int main(void) {                             
    float a[4] = {0.426352024078, 10.556936264038, 29.010599136353, 13.252419471741};     
    float b[4] = {0.426352024078, 10.556936264038, 36.156661987305, 13.252419471741};     
    float threshold = 0.8;                   

    float left = MAX(a[0], b[0]), right = MIN(a[2], b[2]);                                
    float top = MAX(a[1], b[1]), bottom = MIN(a[3], b[3]);                                
    float width = MAX(right - left, (float)0), height = MAX(bottom - top, (float)0);      
    float interS = width * height;           
    float Sa = (a[2] - a[0]) * (a[3] - a[1]);                                             
    float Sb = (b[2] - b[0]) * (b[3] - b[1]);                                             

    int devIoU_ret = interS > threshold * (Sa + Sb - interS); /* devIoU */                

    float ovr = interS / (Sa + Sb - interS); 

    printf("devIoU:         %d\n", devIoU_ret);                                           
    printf("nms_cpu_kernel: %d\n", ovr > threshold);                                      
}
```

And:

```
$ ./a.out
devIoU:         0
nms_cpu_kernel: 1
```

Note this is running entirely on the CPU. The fact that CUDA's involved in the test here appears to be a red herring.

I _think_ the `nms_cpu_kernel()` implementation is the better one, with the rationale being that the operands to the division are more likely to be around the same magnitude (vs the threshold and area coming from 2 different domains in `devIoU()`'s multiplication).

And the `nms_cpu_kernel()` implementation also better matches the python code in `box_iou()` so would presumably be more consistent with that as well.

On the other hand, adjusting `devIoU()` to match `nms_cpu_kernel()` would introduce a division into the CUDA code. I'm not sure whether that's a problem.

Reply:
oh, great catch!

I think adding the division in the cuda kernel is fine, can you fix it?

Reply:
Yes, I'll do some testing and then adjust the PR.

It may be later in the day today; have a few other things competing.

Reply:
Ha! If I add printfs of partial results in those routines, the mismatches go away. Just building without optimizations doesn't seem to prevent the mismatches, though.

Reply:
Thanks a lot for the investigation!

I think for the interest of getting CI green, let's move forward with the mitigation approach you proposed (do not use exact iou), but would be great in the future to get to the bottom of this.